File: package/ios/AdapterSlot/ProviderAdapterSlot.swift
/// Holds a map view's provider adapter and builds a new one only when the
/// configuration it is built from has changed. `HybridMapView` commits to it once
/// per prop transaction.
///
/// `ProviderAdapterSlot.kt` is the Android counterpart; change the two together.
final class ProviderAdapterSlot<Configuration: Equatable, Adapter> {
  private let build: (Configuration) -> Adapter
  private let destroy: (Adapter) -> Void
  private var builtFrom: Configuration?

  private(set) var adapter: Adapter?

  init(
    build: @escaping (Configuration) -> Adapter,
    destroy: @escaping (Adapter) -> Void
  ) {
    self.build = build
    self.destroy = destroy
  }

  /// Makes `adapter` one built from `configuration`: returns it when it had to be
  /// built, and nil when the current one already was. The current adapter is
  /// destroyed before the next one is built, so there are never two at once.
  @discardableResult
  func commit(_ configuration: Configuration) -> Adapter? {
    if adapter != nil, builtFrom == configuration {
      return nil
    }

    release()
    let next = build(configuration)
    adapter = next
    builtFrom = configuration
    return next
  }

  /// Destroys the adapter, if there is one; the next `commit` builds a new one
  /// regardless.
  func release() {
    guard let current = adapter else {
      return
    }

    adapter = nil
    builtFrom = nil
    destroy(current)
  }
}

File: package/android/src/main/java/com/margelo/nitro/nitromaps/ProviderAdapterSlot.kt
package com.margelo.nitro.nitromaps

/**
 * Holds a map view's provider adapter and builds a new one only when the configuration it is
 * built from has changed. [HybridMapView] commits to it once per prop transaction.
 *
 * `ProviderAdapterSlot.swift` is the iOS counterpart; change the two together.
 */
internal class ProviderAdapterSlot<Configuration : Any, Adapter : Any>(
  private val build: (Configuration) -> Adapter,
  private val destroy: (Adapter) -> Unit,
) {
  /** Written on the main thread, read from the JS thread by the imperative methods. */
  @Volatile
  var adapter: Adapter? = null
    private set

  private var builtFrom: Configuration? = null

  /**
   * Makes [adapter] one built from [configuration]: returns it when it had to be built, and null
   * when the current one already was. The current adapter is destroyed before the next one is
   * built, so there are never two at once; a [build] that throws leaves none.
   */
  fun commit(configuration: Configuration): Adapter? {
    if (adapter != null && configuration == builtFrom) {
      return null
    }

    release()
    val next = build(configuration)
    adapter = next
    builtFrom = configuration
    return next
  }

  /** Destroys the adapter, if there is one; the next [commit] builds a new one regardless. */
  fun release() {
    val current = adapter ?: return
    adapter = null
    builtFrom = null
    destroy(current)
  }
}
